### PR TITLE
Request for Code Review

### PR DESCRIPTION
**Pull request by sankalpsharmaa**
Creating this pull request for a code review by @tobylunt for the init branch. 

Looking forward to the feedback and incorporating any changes as necessary.

cc: @seasher04 @paulnov

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,104 @@
...
+
+This code was tested using Stata 16.0. 
+
+---
+# Data Description: SHRUG panel and SHRUG keys
+------
+
+### A)  Data Sources 
+The data comes from 1990, 1998, 2005, and 2013 Economic Censuses (3rd through 6th rounds) and 1991, 2001, 2011 Population Censuses and Village/Town Directories reported by the Indian government.
+
```
Comment (tobylunt) [inline README.md:36]: Need to add a sentence about the Night Lights data as well, since this is an essential component of the paper. Please add:
> Night Lights data used in paper analysis are annual composites of remotely sensed data from the Defense Meteorological Satellite Program Operational Linescan System (DMSP-OLS) at 30 arc second resolution.

File: code/make_shrug_paper.do [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/*****************************************************/
+/* SHRUG  DATA BUILD AND ANALYSIS FOR REPLICATION    */
+/*****************************************************/
+
```
Comment (tobylunt) [inline code/make_shrug_paper.do:6]: Add a comment here to confirm to users that this is the place to start:
> /* START HERE. THIS IS THE MASTER MAKEFILE - MODIFY PATHS BELOW THEN RUN */

File: code/make_shrug_paper.do [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+if "$tmp" == "" | regexm("$tmp", "iec|ddl") ///
+    display  "error: Global tmp not set properly. See instructions in README"
+
+/* set the makefile to crash immediately if globals aren't set properly  */
+if "$out" == "" | regexm("$out", "iec|ddl") ///
+    | "$repdata" == "" | regexm("$repdata", "iec|ddl") ///
+    | "$tmp" == "" | regexm("$tmp", "iec|ddl") ///
+    | "$shcode" == "" | regexm("$shcode", "iec") ///
+    exit 1
```
Comment (tobylunt) [inline code/make_shrug_paper.do:51]: I like this!

File: code/make_shrug_paper.do [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+/************/
+/* analysis */
+/************/
+
+/* anova table */
+do $shcode/a/anova.do
+
+/* table of SECC and IHDS small-area-estimate asset comparison */
+
```
Comment (tobylunt) [inline code/make_shrug_paper.do]: kill empty line

File: tex/front_matter.tex [hunk 5]
```diff
@@ -0,0 +1,65 @@
+\usepackage[latin1]{inputenc}
+% \usepackage{lmodern} % keep or kill this??  might affect italics.
```
Comment (tobylunt) [inline tex/front_matter.tex]: Please delete commented out lines

File: tex/front_matter.tex [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+\usepackage{multirow}
+\usepackage{array}
+\usepackage{harvard}
+\usepackage{tabularx}
+\citationmode{abbr}
+
+\usepackage{float}
+% \usepackage{perpage}
+% \MakeSorted{figure}
+% \MakeSorted{table}
```
Comment (tobylunt) [inline tex/front_matter.tex]: As above - delete

File: code/b/prep_cons_comparison.do [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+save $tmp/ihds_cons_dist, replace
+
+/*********************************************/
+/* prepare district-level NSS for comparison */
+/*********************************************/
+
+/* use schedule 10 to create a household -> state/district key for NSS */
+
+/* save urban district ids in a temporary file */
+use ~/iec1/misc_data/nss68/nss_sch10_urban.dta, clear
```
Comment (tobylunt) [inline code/b/prep_cons_comparison.do:95]: We have a bunch of references to `~/iec1` in this file, which is an error. All data files will be referenced using the globals defined in the makefile - so these `use` statements will fail. Please correct all of these instances in this file, and then do a `grep` within the replication codebase to look for any other instances of `iec*` and `secc` in the repository to be sure there are no others (except for in the comments and checks in the makefile and README.md).

File: code/a/table_sae_decomp.do [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+  cap file close fh
+  
+  /* get the variable list without the constant */
+  local tmp constant
+  local imputevars: list varlist - tmp
+  
+  /********************************************/
+  /* open IHDS and prepare for mean reporting */
+  /********************************************/
+  use $ihds/sae/all_ihds_data, clear
```
Comment (tobylunt) [inline code/a/table_sae_decomp.do]: This `$ihds` global was not set in the makefile so this line will fail. This should be changed to utilize the replication data global

File: code/a/anova.do [hunk 9]
```diff
@@ -0,0 +1,75 @@
+/**************************/
+/* Variance decomposition */
+/**************************/
+
+/* prep dataset */
+clear
+useshrug
+get_shrug_var ec13_emp_all ec13_emp_f total_light_cal2013 avg_forest2014 pc11_pca_tot_work_p pc01_sector pc11_pca_p_06 pc11_pca_tot_work_f pc01_vd_area pc01_td_area pc01_pca_tot_p pc11_sector
+get_shrug_key pc11_district_id pc11_state_id pc11_subdistrict_id
```
Comment (tobylunt) [inline code/a/anova.do:9]: `useshrug` and `get_shrug_var` are tools that could cause problems here. 
1. They're located in `~/ddl/tools/do/get_vars.do` and may not have been added to this replication repo's programs
2. They depend on paths that will not work outside of DDL's Polaris environment. E.g. `get_shrug_var` depends on `~/iec/../shrug_varlist.txt`, which the replicator will not have.
This is why we need to test replication repositories with the test user account

File: code/a/anova.do [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+drop pc01_vd_area_convert
+
+/* prepare calculated vars */
+gen ec13_emp_f_share = ec13_emp_f / ec13_emp_all 
+gen ec13_emp_pc = ec13_emp_all / (pc11_pca_tot_p - (15 * pc11_pca_p_06 / 7))
+gen nl_pc = total_light_cal_2013 / pc11_pca_tot_p
+gen f_lab_shr = pc11_pca_tot_work_f / pc11_pca_tot_work_p
+gen density = pc01_pca_tot_p / pc01_area
+
+/* TMP: fix errors in consumption variable */
```
Comment (tobylunt) [inline code/a/anova.do]: remove `TMP:` from this comment

File: README.md (hunk 1 above)
Comment (tobylunt) [reply README.md:36]: Also add:
> Consumption per capita comparisons in this paper also utilize data from the National Sample Survey (NSS) and India Human Development Survey (IHDS).

File: code/a/analyze_firms.do [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+  
+  binscatter `1' `2' `if', linetype(none) name(`n', replace) ///
+      `control' `xscale' `yscale' `xlabel' `ylabel' `xtitle' `ytitle' `title' xq(__xcut)
+  graphout `n'
+  capdrop __rank __xcut
+end
+/* *********** END program rbin ***************************************** */
+
+
+/* prepare analysis dataset */
```
Comment (tobylunt) [inline code/a/analyze_firms.do:87]: This would be better as a comment box rather than a single line comment

File: code/a/analyze_firms.do [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+    /* append stats to an output file */
+    foreach v in p20 p50 p80 {
+      append_to_file using $f, s("`locname`loc''_`s'_`v',``v''")
+    }
+    append_to_file using $f, s("`locname`loc''_`s'_mean,`mean'")
+    append_to_file using $f, s("`locname`loc''_`s'_beta,`b'")
+    append_to_file using $f, s("`locname`loc''_`s'_betadist,`bd'")
+  }
+}
+cat $f
```
Comment (tobylunt) [inline code/a/analyze_firms.do:215]: Let it be known that the code style in this section is not up to DDL standards - it's inadequately commented with inconsistent linebreaks. But leave it as-is so we can get this out.

File: code/a/analyze_nl_boot.do [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+  noi di `b'
+  
+  /* set consumption to the current bootstrap value */
+  replace ln_consumption = ln(secc_cons_pc_`b')
+  
+  /* run the regressions and store the bootstrap coef */
+  quireg ln_consumption ln_light2012 ln_num_cells, robust title(ln_consumption) absorb(sdgroup)
+  append_est_to_file using $tmpfile, b(ln_light2012) s(txc_d_cons,`b') 
+
```
Comment (tobylunt) [inline code/a/analyze_nl_boot.do]: delete this blank line

File: code/a/analyze_nl_boot.do [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+  local stars `r(stars)'
+  local starbeta `beta'`stars'
+  
+  /* insert the stats into the output file */
+  insert_into_file using $output_csv, key(`stat'_cons_starbeta) val("`starbeta'")
+  insert_into_file using $output_csv, key(`stat'_cons_se) val(`se')
+}
+
+shell diff $out/nl_xc.csv $output_csv
+
```
Comment (tobylunt) [inline code/a/analyze_nl_boot.do]: you can delete the `diff` - this was an internal check

File: code/a/analyze_poverty.do [hunk 15]
```diff
@@ -0,0 +1,180 @@
+/* GOAL: Create SHRUG rural poverty analysis dataset */
+
+/* get rural poverty rate and consumption level */
+use shrid secc_pov_rate_tend_rural using $shrug/data/shrug_secc, clear
+ren secc_pov_rate_tend_rural secc_pov_rate_rural
+
+/* get district and subdistrict ids */
+merge 1:1 shrid using $shrug/keys/shrug_pc11_district_key, keepusing(pc11_state_id pc11_district_id) nogen keep(match master)
+merge 1:m shrid using $shrug/keys/shrug_pc11_subdistrict_key, keepusing(pc11_subdistrict_id) nogen keep(match master)
```
Comment (tobylunt) [inline code/a/analyze_poverty.do:9]: Note that `$shrug/keys` and `$shrug/data` paths will need to match exactly in the replication data folder. Seeing that this repo hasn't been tested with the test user yet, this is something that may break.

File: code/a/validate_con_imputation.do [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+3. compare the totals to $shrug/intermediate/con_pre_impute_08.
+- mean absolute deviation as a percentage of constituency population
+
+4. graph: share shrids dropped on X axis, mean absolute deviation on Y axis.
+
+*/
+
+/**********************************************************************************************************************/
+/* program impute_shrid_pop : Impute population for missing villages based on other villages in the same constituency */
+/*   note: this is copy-pasted from collapse_shrug_to_con.do exactly */
```
Comment (tobylunt) [inline code/a/validate_con_imputation.do]: replicators don't know what `collapse_shrug_to_con.do` is since it's not in the replication repo. delete this line

File: code/a/validate_con_imputation.do [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+// merge m:1 pc11_state_id using $tmp/mean_con_pop, keep(match master) keepusing(mean_con_pop)
+// 
+// /* get total constituency population based on shrid match */
+// bys ac08_id: egen con_pop = total(pc11_pca_tot_p)
+// tag ac08_id
+// gen con_pop_ratio = con_pop / mean_con_pop
+// assert !mi(con_pop_ratio)
+// drop if !inrange(con_pop_ratio, .4, 2.5)
+// drop con_pop con_pop_ratio mean_con_pop 
+
```
Comment (tobylunt) [inline code/a/validate_con_imputation.do]: delete

File: code/a/validate_con_imputation.do [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+qui forval threshold = .01(.01).30 { 
+  noi di "Test: setting `threshold' observations to missing..."
+  local threshstr = round(`threshold' * 100)
+
+  use $tmp/pre_impute_test, clear
+
+  /* set a random subset of pc01 populations to missing */
+  foreach loc in "_r" "_u" {
+    //   replace pc01_pca_tot_p_r = . if runiform() < `threshold'
+    //   replace pc01_pca_tot_p_u = . if runiform() < `threshold'
```
Comment (tobylunt) [inline code/a/validate_con_imputation.do]: delete

File: code/a/validate_con_imputation.do [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+  impute_shrid_pop, group_id(ac08_id) base_year(11) target_year(01) pca_suffix("")
+
+  /* drop intermediate variables used in population imputation */
+  keep shrid ac08_id pc* c_overlap_ratio c_misaligned_pop_share c_pop* pop_old
+  drop c_pop*found*
+
+  /* compare the result to the original file with no drops */
+  ren pc01_pca_tot_p pop01_imputed
+  // ren pc01_pca_tot_p_r pop01_imputed_r
+  // ren pc01_pca_tot_p_u pop01_imputed_u
```
Comment (tobylunt) [inline code/a/validate_con_imputation.do]: delete

File: code/a/validate_con_imputation.do [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+  foreach loc in "_r" "_u" {
+    //   replace pc01_pca_tot_p_r = . if runiform() < `threshold'
+    //   replace pc01_pca_tot_p_u = . if runiform() < `threshold'
+    replace pc01_pca_tot_p = . if runiform() < `threshold'
+  }
+
+  /* impute missing all missing 2001 populations */
+  gen pop_old = pc01_pca_tot_p 
+  // impute_shrid_pop, group_id(ac08_id) base_year(11) target_year(01) pca_suffix(_r)
+  // impute_shrid_pop, group_id(ac08_id) base_year(11) target_year(01) pca_suffix(_u)
```
Comment (tobylunt) [inline code/a/validate_con_imputation.do]: delete

File: code/ado/tools.do [hunk 21]
```diff
@@ -0,0 +1,5462 @@
+qui {
```
Comment (tobylunt) [inline code/ado/tools.do:1]: We should not be including all of `tools.do` in the replication repo.

Please follow instructions [here](https://github.com/devdatalab/ddl-manual/wiki/replication-repos#find-toolsdo-programs-used-in-repo-grep_include_progssh)

File: code/make_shrug_paper.do [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+if "$out" == "" | regexm("$out", "iec|ddl") ///
+    | "$repdata" == "" | regexm("$repdata", "iec|ddl") ///
+    | "$tmp" == "" | regexm("$tmp", "iec|ddl") ///
+    | "$shcode" == "" | regexm("$shcode", "iec") ///
+    exit 1
+
+/***************/
+/* preparation */
+/***************/
+
```
Comment (tobylunt) [inline code/make_shrug_paper.do:60]: General note: please double check that all scripts listed in the makefile are included in the correct location (this implicitly happens when you run the script with the test user account) and that there are not extraneous scripts that aren't called (sometimes secondarily) by the makefile

**Review by tobylunt — CHANGES_REQUESTED** [review]
# General comments

Overall, nice work. This is a tricky job, and this is a great start.

Specifics:
* Delete all PDF outputs from the repository. These will be generated by the user when they run the code
* Where are all the table templates? These are the `.tpl` files that build out the final Latex tables and are required for the repo to work
* Delete Latex tables that are final output from the repo. Like the PDF output, Latex table outputs created by the code should not be shipped in the repo - they will be created by the user in the code, and depend on the table templates (unless there are any pre-made Latex tables that aren't programmatically created, those should be included in the repo).
    * More explicitly: the only files in `tex/` should be those that are not autogenerated, like the paper, paper appendix, `master.bib`, `table_shrug_contents`, `front_matter`, maybe others
    
# Biggest comment

It's clear that the replication code hasn't been run using the test user account, because there are a bunch of paths that would throw errors and break the code. So - please do this! [Instructions on how to test a replication repo are here](https://github.com/devdatalab/ddl-manual/wiki/replication-repos#f-test-the-replication-repo).

As you test, you might run into issues that you can't immediately understand. Please don't hesitate to reach out to Toby or Becky right away, they may have a quick answer that saves you time.

# Next steps

Basically, two things:

1. Go through and make the fixes I've requested. Submit another code review request and we'll iterate as needed until everything looks great
2. Once everything looks great, we actually will not merge the PR to the master branch, because we prefer not to retain the commit history ([as described here](https://github.com/devdatalab/ddl-manual/wiki/replication-repos#g-clean-up-the-branches)). So instead, you'll copy the code to the other branch and then delete the `init` branch.

File: README.md (hunk 1 above)
Comment (sankalpsharmaa) [reply README.md:36]: Added in this [commit](https://github.com/devdatalab/paper-shrug/commit/9884cdc67a0bce88487dd20b3483ba915905bf30)

File: code/make_shrug_paper.do (hunk 2 above)
Comment (sankalpsharmaa) [reply code/make_shrug_paper.do:6]: Added in this [commit](https://github.com/devdatalab/paper-shrug/commit/8e3fcf0973f14d5e81896389c38b4d9e8fbebd71)

File: code/make_shrug_paper.do (hunk 4 above)
Comment (sankalpsharmaa) [reply code/make_shrug_paper.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/dd83f3fa5ec7cb480417671e185e07a0223f0333)

File: tex/front_matter.tex (hunk 5 above)
Comment (sankalpsharmaa) [reply tex/front_matter.tex]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/688afe7f403dd7b0619b0b0d2b2c8334c245c15d)

File: tex/front_matter.tex (hunk 6 above)
Comment (sankalpsharmaa) [reply tex/front_matter.tex]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/1fb806e7261d20e59cdb6c4efb1625c991ec6208)

File: code/a/anova.do (hunk 10 above)
Comment (sankalpsharmaa) [reply code/a/anova.do]: fixed [here](https://github.com/devdatalab/paper-shrug/commit/ee3efada7b4101a88aaea39775b975e17edecc3b)

File: code/a/analyze_firms.do (hunk 12 above)
Comment (sankalpsharmaa) [reply code/a/analyze_firms.do:215]: Should I add the necessary comments?

File: code/a/analyze_nl_boot.do (hunk 13 above)
Comment (sankalpsharmaa) [reply code/a/analyze_nl_boot.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/ac6b4450ab5172fc6e95e16118c4a0b0ffefd180)

File: code/a/analyze_nl_boot.do (hunk 14 above)
Comment (sankalpsharmaa) [reply code/a/analyze_nl_boot.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/ac6b4450ab5172fc6e95e16118c4a0b0ffefd180)

File: code/a/validate_con_imputation.do (hunk 17 above)
Comment (sankalpsharmaa) [reply code/a/validate_con_imputation.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/19a059b092ad2fa54f49dccc7f3808daae340eb1)

File: code/a/validate_con_imputation.do (hunk 18 above)
Comment (sankalpsharmaa) [reply code/a/validate_con_imputation.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/19a059b092ad2fa54f49dccc7f3808daae340eb1)

File: code/a/validate_con_imputation.do (hunk 19 above)
Comment (sankalpsharmaa) [reply code/a/validate_con_imputation.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/19a059b092ad2fa54f49dccc7f3808daae340eb1)

File: code/a/validate_con_imputation.do (hunk 20 above)
Comment (sankalpsharmaa) [reply code/a/validate_con_imputation.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/19a059b092ad2fa54f49dccc7f3808daae340eb1)

File: code/a/validate_con_imputation.do (hunk 16 above)
Comment (sankalpsharmaa) [reply code/a/validate_con_imputation.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/19a059b092ad2fa54f49dccc7f3808daae340eb1)

File: code/a/table_sae_decomp.do (hunk 8 above)
Comment (sankalpsharmaa) [reply code/a/table_sae_decomp.do]: fixed in this [commit](https://github.com/devdatalab/paper-shrug/commit/31a545ea89f35773eb24110ab3062949223e8ba0)

File: code/a/analyze_poverty.do (hunk 15 above)
Comment (sankalpsharmaa) [reply code/a/analyze_poverty.do:9]: the `shrug` global has been set to `repdata` so hopefully this shouldn't create a problem. Will update this once I run the makefile on the test user account.

File: code/a/analyze_firms.do (hunk 12 above)
Comment (tobylunt) [reply code/a/analyze_firms.do:215]: no, please leave as-is

**Review by tobylunt — APPROVED** [review]
Looks like you got the changes I requested so I'm merging the branches. I will do a final code review once the replication repo has been tested with the test user account.
